### ISSUE TITLE
Regression: Navigation triggered outside Angular zone

### ISSUE THREAD

**Issue by Eugeno** [ag-grid/ag-grid#7464]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of AG Grid'
[] support request => see 'Requesting Community Support'
```

**Providing a Reproducible Scenario**
```html
<a [routerLink]="[params.path, params.value]">Link</a>
```
When we follow the link, we get a warning `Navigation triggered outside Angular zone, did you forget to call 'ngZone.run()'?` and `ngOnInit()` doesn't fire until we click somewhere.

This bug has appeared in version **31.0.2**, everything is fine in versions 31.0.1 and earlier.
Angular version: 17.1.0.

Temporal workaround (if anyone else is facing the same problem):
```html
<a href="javascript:void(0)" (click)="navigate([params.path, params.value])">Link</a>
```
```ts
navigate(parts: unknown[]): void {
  this.ngZone.run(() => this.router.navigate(parts));
}
```

**Comment by StephenCooper** [comment]
Hi @Eugeno,

Could you provide a bit more information about where that link is? Is this within a cell renderer?

**Comment by Eugeno** [comment]
@StephenCooper oh, sorry about that.
Yes, this is in cellRenderer component (standalone, ChangeDetectionStrategy.OnPush).

**Comment by StephenCooper** [comment]
Are you able to share the code of the cell renderer?

**Comment by Eugeno** [comment]
I'm not sure if this is allowed by the NDA. But if we set aside the business logic, there's just a link there, where its parameters are passed through cellRendererParams. Perhaps, later on, I'll find time to create an example on Stackblitz.

**Comment by StephenCooper** [comment]
I have tried to reproduce with the following renderer but not able to see the warning. 

```
// Custom Cell Renderer Component
@Component({
  standalone: true,
  imports: [RouterLink],
  template: `
    <a [routerLink]="['/second-component']">Link</a>
  `,
})
export class LinkRenderer implements ICellRendererAngularComp {
  // Init Cell Value
  public params!: ICellRendererParams;
  agInit(params: ICellRendererParams): void {
    this.params = params; 
  }

  // Return Cell Value
  refresh(params: ICellRendererParams): boolean {
    this.params = params;
    return true;
  }
}
```

**Comment by Eugeno** [comment]
I also decided to create a basic StackBlitz, and it didn't reproduce there either. However, on a real project, when I roll back to version 31.0.1, it starts working.
Well, it seems there is some deeper problem.
I wanted to see what changed in version 31.0.2, but in the Diffs, there are a million lines 😅

**Comment by Eugeno** [comment]
Oh, I found it! It appears when link has condition (we check role permissions):
```html
@if (hasPermission) {
  <a [routerLink]="['test']">Link</a>
}
```
It's not related to the new control flow (@&#8202;if), same problem with *ngIf.

Here are the repo:
https://stackblitz-starters-uhqcx3.stackblitz.io

**Comment by StephenCooper** [comment]
Thank you for that. It has help me identify the source of the issue and I believe I can resolve it.

**Comment by StephenCooper** [comment]
The fix for this issue is now with QA, with the aim of being part of a patch release sometime next week.

I have updated the ZoneJs tests to include this use case which unfortunately was missed so this shouldn't happen again. 

Thank you very much for your help in pinpointing the issue which made it quicker for me to resolve.

**Comment by AG-Zoheil** [comment]
Now that this requirement is recorded in our backlog, we will close the GitHub issue because we will not update it when this feature is added in a future version of AG Grid. The best way to get an update on this item is to sign up for AG Grid new release notifications. Once a new AG Grid version is released, look for the item reference above (AG-10405) on the changelog page to see if it's resolved in that version:
https://www.ag-grid.com/ag-grid-changelog/

Thanks again for bringing this up with us!

**Comment by Eugeno** [comment]
@StephenCooper I'm not entirely sure, it's the same issue, but it also appeared in version 31.0.2 and is related to cellRenderer. Some details: the component with a cellRenderer renders fine, but if we open a modal window, which also contains a table with the same cellRenderer, we get an Angular error: `Uncaught TypeError: Cannot read properties of undefined (reading 'ɵcmp')`.
I managed to create a [minimal reproducible repository](https://github.com/Eugeno/ag-grid-error).
Let me know if it's a different problem, and if so, I'll create a new issue.

**Comment by Eugeno** [comment]
@StephenCooper I'm not entirely sure, it's the same issue, but it also appeared in version 31.0.2 and is related to cellRenderer, although the error itself is completely different.
Some details: the component with a cellRenderer renders fine, but if we open a modal window, which also contains a table with the same cellRenderer, we get an Angular error: `Uncaught TypeError: Cannot read properties of undefined (reading 'ɵcmp')`.
I managed to create a [minimal reproducible repository](https://github.com/Eugeno/ag-grid-error).
Could you check if your fix covers this case? If yes, then I'll just wait for the new version; if not, I can create a new issue.

**Comment by StephenCooper** [comment]
> @StephenCooper I'm not entirely sure, it's the same issue, but it also appeared in version 31.0.2 and is related to cellRenderer, although the error itself is completely different. Some details: the component with a cellRenderer renders fine, but if we open a modal window, which also contains a table with the same cellRenderer, we get an Angular error: `Uncaught TypeError: Cannot read properties of undefined (reading 'ɵcmp')`. I managed to create a [minimal reproducible repository](https://github.com/Eugeno/ag-grid-error). Could you check if your fix covers this case? If yes, then I'll just wait for the new version; if not, I can create a new issue.

Which version where you using previously as it seems like this error happens even with older versions. Is it just AG Grid that that been updated in this repo?

**Comment by Eugeno** [comment]
Yes, indeed, this bug reproduces even in earlier versions. I apologize for not checking. So, it should definitely be a separate issue. I'll try to figure out why similar code worked for me in the past.

**Comment by Eugeno** [comment]
Yes, indeed, this bug reproduces even in earlier versions (used 31.0.1). I apologize for not checking. So, it should definitely be a separate issue. I'll try to figure out why similar code worked for me in the past.

**Comment by Eugeno** [comment]
Well, in my case, it happened when I removed lazy loading of the component:
```ts
import type { DialogComponent } from '../dialog';

const { DialogComponent } = await import('../dialog');
this.dialogService.open(DialogComponent);
```
replaced with:
```ts
import { DialogComponent } from '../dialog';

this.dialogService.open(DialogComponent);
```
(replacement was made to fix one of the reactivity problems for current issue).

There used to be no lazy loading, and it worked, but since then, there have been many releases of both ag-grid and angular, and it's hard to trace back.
Btw, in the [test repository](https://github.com/Eugeno/ag-grid-error), I removed the use of nebular to eliminate this influencing factor.

**Comment by urrang** [comment]
We're also seeing some issues that feels related to this. I haven't been able to reproduce outside of our app yet, but I'll add some info about the issues in hopes that it can help identify what's going on.

The issues appeared when upgrading from 31.0.1 to 31.0.2, and do not seem to be fixed in 31.0.3.

They seem to be related to components that are dynamically loaded (using e.g `createComponent(...)`) from inside our grid component. A few examples:

- We have a dialog service that dynamically creates dialog components and attaches them to a portal. Calling `showDialog(..)` outside our grid component is basically instant, but if the call happens inside e.g a cell renderer then it takes anywhere from 500ms to 2 seconds for the dialog to appear. No idea why, but using `console.time` most of the delay seems to be between `constructor()` and `ngOnInit()` in the dialog component.

- We have a custom cell renderer for dropdown menus. The component has a button, and on click it'll dynamically create a dropdown-panel component. The button has `*ngIf="menuItems.length"` to control whether or not it should show on a given row. When clicking the button it takes 500-1500ms before the panel is displayed, however if we remove the ngIf, or run `changeDetectorRef.markForCheck()` at the end of `agInit` (where we set menuItems) the panel shows instantly on click. This makes very little sense to me, as the ngIf functions fine in itself without `markForCheck`, and I don't see why any of this should affect the onClick logic.

- We have a custom typeahead/autocomplete cell editor. The component uses `ChangeDetectionStrategy.onPush`, but we're careful to call `changeDetectorRef.markForCheck()` when necessary. However on input events there's still a noticable delay between new autocomplete items being available and them actually showing up in the view.

**Comment by StephenCooper** [comment]
Thanks for providing more details about this and I would definitely be interested in trying to understand this more. 

Does the Angular dev tools profiler give any insight into differences? 

Is there a difference in whether agInit is running in / out of the zone between 31.0.1 and 30.0.3?

**Comment by urrang** [comment]
Thanks for the quick response!

I went back to my last commit before adding the markForCheck calls, and.. it works just fine on 31.0.3.

I've downgraded to 31.0.2 and there definitely is an issue there, so I'm leaning towards my confusion being a mix of ^version in package.json, and not paying attention to the output of npm install when jumping between versions 🤦 

markForCheck in agInit does make noticeable difference on 31.0.2, though I'm unable to find a good reason why. The function runs inside zone in both 31.0.2 and 31.0.3. I haven't used the Angular profiler much before so I'm not entirely sure what I'm looking at, but there seems to be a long lasting setInterval right after the click, which is I'm unable to find when calling markForCheck in agInit.

```json
{
    ...
    {
        "source": "HTMLElement.addEventListener:click",
        "duration": 0.3999999985098839,
        "directives": [...]
    },
    {
        "source": "setInterval",
        "duration": 1.7000000029802322,
        "directives": [
            {
                "children": [
                    {
                        "children": [
                            {
                                "children": [],
                                "directives": [
                                    {
                                        "isComponent": false,
                                        "isElement": false,
                                        "name": "_NgIf",
                                        "outputs": {},
                                        "lifecycle": {}
                                    }
                                ]
                            },
                            {
                                "children": [],
                                "directives": [
                                    {
                                        "isComponent": false,
                                        "isElement": false,
                                        "name": "_NgForOf",
                                        "outputs": {},
                                        "lifecycle": {}
                                    }
                                ]
                            }
                        ],
                        ...
                    }
                ]
            }
        ]
    },
    ...
}
```

I'd be happy to help if there's anything you want me to test/profile, but seeing as it's working in 31.0.3 that's good enough for me 😄 

Regarding the showDialog function being passed from our grid wrapper (angular component) and called in a regular js cell renderer that's still a bit yanky on 31.0.3. I just checked, and it does indeed run outside ngzone in 31.03 and 31.0.2, and inside ngzone in 31.0.1. I don't really mind this as we had plans to refactor those renderers to Angular components either way, but I thought I'd let you know in case it's unintentional.

**Comment by StephenCooper** [comment]
I am glad to hear the main issues are resolved in v31.0.3. Stale npm versions are always a gotcha! 

The fact that it is running outside in 31.0.3 is expected but the fact that it slows things down is interesting. I wonder if what you are seeing is that calling the dialog while outside of ngZone means it does not get updated until some other part of your application triggers change detection to run.

My guess would be that is you triggered change detection at the end of your showDialog function that this would resolve the yankiness. If you don't refactor those components to Angular components then you may want to wrap the contents of the showDialog function within ngZone.run(). 

Thank you for bring this edge case to my attention though of calling back into Angular from a JS component.

**Comment by svenank3r** [comment]
Hi,

Not totally sure on the status of this issue as it was supposedly fixed in 31.0.3, but I still encounter this issue.

I do not use a cellRenderer or dialog, but I just simply use the Angular router in a service to navigate to a page/route. The only work-around right now is to enclose the call to the service with the ngZone.run() function.

As far as I know this broke since 31.0.2 and the above comment mentions "The fact that it is running outside in 31.0.3 is expected", but I would not expect such a big change in a patch version?
Or am I missing out on some big change that was recently made?

I have created a [StackBlitz](https://stackblitz-starters-lhbijq.stackblitz.io) with my case.

**Comment by StephenCooper** [comment]
Hi @svenank3r,

Thank you for raising this I will look into this. 

In the meantime if you move your `onCellClicked` callback to `gridOptions` instead of the `defaultColDef` then this will work as expected. 

https://stackblitz.com/edit/stackblitz-starters-vjo6u8?file=src%2Fhome.component.ts

I do appreciate that this change has likely been mis-versioned given the scope of its impact.

**Comment by StephenCooper** [comment]
To follow up and set expectations incase others come across this ticket. 

If v31.0.3 does not work for your use case please pin to v31.0.1.  We are not expecting to release any more patches with respect to this issue. However, please do report your issues if there are still any.

We will aim to resolve any reported issues in our next minor version 31.1.0. This will include the fix for column events reported above.
